### PR TITLE
Danielh/eng 383 generate a tech support object from

### PR DESCRIPTION
**Pull request by DanielHHowell**
Some notes:

- Tried to prevent raising in general on the diagnostics ("assets") API, due to user mis-setting or otherwise as this is kind of a siloed beta functionality and we don't want to bring the whole house down
-  Utilizing [`aiofiles`](https://github.com/Tinche/aiofiles) to maintain proper async
- Corrected the `no_poll` implementation in the `checks` command as `no_diagnostics` functions identically

As I haven't really done servox dev outside of external connectors, likely some ample room for improvement

Can be tested with quick and dirty e-tools script `diagnostics.py`

### CODE REVIEW

File: servo/api.py
Comment (linkous8) [inline servo/api.py:81]: Please move `Diagnostics` to servo/telemetry.py

File: servo/api.py
Comment (linkous8) [inline servo/api.py:318]: There's a lot of repetition between the methods `_diagnostics_request`, `_post_diagnostics`, and `_reset_diagnostics`. Please refactor the common code into a helper method in servo/telemetry.py

File: servo/runner.py
Comment (linkous8) [inline servo/runner.py:224]: Please move to a new class (eg. DiagnosticsChecker) in servo/telemetry.py

File: servo/runner.py
Comment (linkous8) [inline servo/runner.py:274]: Please move diagnostic task init to the `run` method just above this line https://github.com/opsani/servox/blob/e5d744cb298f6a93d1c55354ba6300f246be67e8/servo/runner.py#L325

File: servo/runner.py
Comment (linkous8) [inline servo/runner.py:276]: Diagnostics task should log errors and not re-raise (please add exponential backoff and/or self-termination if there are repeated errors to prevent logs from getting filled with repeat errors)

File: servo/cli.py
Comment (linkous8) [inline servo/cli.py:1252]: Please replace this with a new option on `servo.configuration.BaseServoConfiguration`. Analogous shell invocation would be `SERVO_NO_DIAGNOSTICS=True servo run ....`

File: servo/cli.py
Comment (linkous8) [inline servo/cli.py:1158]: Please replace this with a new option on `servo.configuration.BaseServoConfiguration`. Analogous shell invocation would be `SERVO_NO_DIAGNOSTICS=True servo run ....`

**Review by linkous8 — CHANGES_REQUESTED** [review]
In addition to the inline change requests, please move the following to a new class (or classes) in servo/telemetry.py (eg. DiagnosticsChecker). I recommend having it inherit servo.api.Mixin so it has access to the api_client as a member:
* Diagnostics
* _get_diagnostics
* _diagnostics_request
* _post_diagnostics
* _reset_diagnostics
* def running_diagnostics
* async def diagnostics_check

Further, please remove the added `diagnostics` arguments from the various run methods by adding a check on the config here (eg. `self.servo.config.no_diagnostics`): https://github.com/opsani/servox/blob/e5d744cb298f6a93d1c55354ba6300f246be67e8/servo/runner.py#L325

File: servo/runner.py
Comment (DanielHHowell) [reply servo/runner.py:224]: Ok, I figured it was a part of the API but makes sense to silo. I am having trouble understanding how the Diagnostics class should access the running servo however. If I make a `class DiagnosticsHandler(servo.api.Mixin)` in servo/telemetry.py, what is the correct way for it to operate on the running servo within the runner?

File: servo/api.py
Comment (DanielHHowell) [reply servo/api.py:318]: Good call

File: servo/runner.py
Comment (linkous8) [reply servo/runner.py:224]: You can get the servo for the current context (usually only one context/servo per assembly anyway) by using the `servo.current_servo` method https://github.com/opsani/servox/blob/e5d744cb298f6a93d1c55354ba6300f246be67e8/servo/servo.py#L29

Also note its okay for the runner to continue managing the repeating task (since the runner also [sets up cancellation handling](https://github.com/opsani/servox/blob/e5d744cb298f6a93d1c55354ba6300f246be67e8/servo/runner.py#L393) and needs to be aware of the diagnostics task so its not stopped indefinitely (or at all)). The code run by the task should be part of telemetry.py and the runner is responsible for init (and, optionally, calling a simple start method if it makes sense to implement one outside of the init)

File: servo/cli.py
Comment (DanielHHowell) [reply servo/cli.py:1158]: Ok, I think I've got everything in except this one - likely in need of an implementation hint here

Comment (linkous8) [reply servo/cli.py:1158]: Add a new field here: https://github.com/opsani/servox/blob/138f762965073a865b1742f2c9e448d273e2eff3/servo/configuration.py#L491

```
no_diagnostics: bool = pydantic.Field(
    default=False, description="Do not poll the Opsani API for diagnostics"
)
```

It can then be referenced on the configuration of the servo `current_servo.config.no_diagnostics`

(also feel free to invert the name of this if you think it makes more sense `no_diagnostic` -> `diagnostics`)

File: servo/telemetry.py
Comment (linkous8) [inline servo/telemetry.py:109]:
```suggestion
                    self.servo.config.no_diagnostics = True
                    self._running = False
```

File: servo/telemetry.py
Comment (linkous8) [inline servo/telemetry.py:105]:
```suggestion
                    await self._put_diagnostics(diagnostic_data)
                    await self._reset_diagnostics()
```

File: servo/telemetry.py
Comment (linkous8) [inline servo/telemetry.py:94]:
```suggestion
```

File: tests/api_test.py
Comment (linkous8) [inline tests/api_test.py:3]: Please remove unused imports
```suggestion
import pytest
```

**Review by linkous8 — CHANGES_REQUESTED** [review]

File: servo/runner.py
Comment (DanielHHowell) [reply servo/runner.py:296]: Hmm, it was there originally but moved it to the run based on [first review comment](https://github.com/opsani/servox/pull/373#discussion_r766001691)? What do you think is best?

Comment (linkous8) [reply servo/runner.py:296]: You're right, my bad. I think it can be left where it is and instead just remove the task cancellation as part of the main loop startup (or move the cancellation to where the task is being created currently) https://github.com/opsani/servox/blob/ff9ec7b1b305037c3afd984ae124e69b5f6a6ef1/servo/runner.py#L226-L227

File: servo/runner.py
Comment (DanielHHowell) [reply servo/runner.py:35]: Ah, interesting. I've recently switched to VSCode and the GitLens line comments had me falsely believe one to be already present

File: servo/runner.py
Comment (DanielHHowell) [reply servo/runner.py:296]: I do actually prefer it where it was originally so both loops are on the same tier

File: servo/telemetry.py
Comment (DanielHHowell) [reply servo/telemetry.py:30]: The [backoff decorator](https://github.com/litl/backoff#logging-configuration) logs to `INFO` by default and I don't want the logs to crowd with retry attempts for diagnostics at the info level

File: servo/telemetry.py
Comment (DanielHHowell) [reply servo/telemetry.py:140]: I tried that, but there were some that did not convert succesfully. In the CLI method Blake had these roundtripped for I imagine the same purpose

File: servo/telemetry.py
Comment (DanielHHowell) [reply servo/telemetry.py:208]: I marinated on that feedback from the first review, and did extract a helper method for the client.puts() to use, but the response reading and return function differently for `request`, I feel like readability would be reduced.

File: servo/telemetry.py
Comment (linkous8) [reply servo/telemetry.py:30]: Good thinking. Only issue is you're increasing the verbosity instead of decreasing it: https://loguru.readthedocs.io/en/stable/api/logger.html#levels

I suggest setting to the DEBUG level (also are you sure its getting the same sinks as the main logging assembly so that it gets logged to stdout and the servo.log file?)

File: servo/telemetry.py
Comment (linkous8) [reply servo/telemetry.py:207]: Use the `request` method with arguments for put and get https://www.python-httpx.org/api/

The error handling is largely the same so just inject the parsing object as a paremeter (the model or a lambda)

```
    async def _diagnostics_request(
        self, 
        method: str, 
        endpoint: str, 
        output_model: pydantic.BaseModel,
        json: Optional[dict] = None,
    ) -> Union[DiagnosticStates, Status]:
        async with self.api_client() as client:
            self.logger.trace(f"{method} diagnostic request")
            try:
                response = await client.request(method=method, url=endpoint, json=json)
                response.raise_for_status()
                response_json = response.json()
                self.logger.trace(
                    f"{method} diagnostic request response ({response.status_code} {response.reason_phrase}): {devtools.pformat(response_json)}"
                )
                self.logger.trace(servo.api._redacted_to_curl(response.request))
                try:
                    return pydantic.parse_obj_as(
                        output_model, response_json
                    )
                except pydantic.ValidationError as error:
                    # Should not raise due to improperly set diagnostic states
                    self.logger.exception(f"Malformed diagnostic {method} response", level_id="DEBUG")
                    return DiagnosticStates.withhold

            except httpx.HTTPError as error:
                self.logger.trace(servo.api._redacted_to_curl(error.request))
                raise
```

File: servo/telemetry.py
Comment (linkous8) [reply servo/telemetry.py:139]: In that case, I think we can just drop this model conversion alltogether and just add the servo config to the diagnostics object. Then it should be serialized into json by httpx's json parameter

File: servo/telemetry.py
Comment (DanielHHowell) [reply servo/telemetry.py:30]: Good catch on adding the InterceptHandler. As for the level, they actually say "if you would instead like to log only when a giveup event occurs, set the logger level to ERROR." Tested to verify~

File: servo/telemetry.py
Comment (DanielHHowell) [reply servo/telemetry.py:139]: I don't like the idea of not pushing it through the exclude filters as one is wading through thick None and Null fields otherwise

Comment (linkous8) [reply servo/telemetry.py:145]: Good point. Further, the model needs to be converted to json by calling pydantic's `.json()` method but httpx will always call `json.dumps()`. Pydantic's json can serialise many commonly used types to JSON (e.g. datetime, date or UUID) which would normally fail json.dumps but theres no way to ensure httpx will use that json encoding implementation (see https://github.com/samuelcolvin/pydantic/issues/951)

Maybe add a todo comment to reevaluate the json round tripping logic later down the road

File: servo/telemetry.py
Comment (linkous8) [reply servo/telemetry.py:29]: Okay thanks for clarifying, I misunderstood how `setLevel` was working; I thought it was setting the log level of logs emitted by backoff but its actually configuring the minimum severity threshold for backoff's logging filter :+1:

**Comment by DanielHHowell** [comment]
@linkous8 all good to merge here after the integration tests finish?

File: tests/telemetry_test.py
Comment (linkous8) [inline tests/telemetry_test.py:47]:
```suggestion
@pytest.mark.parametrize(
    "input, expected_state",
    [
        ("WITHHOLD", servo.telemetry.DiagnosticStates.withhold),
        ("SEND", servo.telemetry.DiagnosticStates.send),
        ("STOP", servo.telemetry.DiagnosticStates.stop),
    ],
)
def test_withhold_diagnostic_state(input, state) -> None:
    state = servo.telemetry.DiagnosticStates(input)
    assert state == expected_state
```

File: tests/telemetry_test.py
Comment (linkous8) [inline tests/telemetry_test.py:59]: Don't need to simulate running in k8s for these tests

```suggestion
async def test_diagnostics_request(servo_runner: servo.runner.ServoRunner, optimizer: servo.configuration.Optimizer) -> None:
```

**Review by linkous8 — APPROVED** [review]
It would be nice to have a test that runs the `diagnostics_check` for a quick end to end verification but I'm not going to block on that. Feel free to implement post 2.5

(Also it may be easier to test that method if you move the while loop to another method to make it easier to run a single check eg. `async def diagnostics_check_loop(self)`  and `async def diagnostics_check(self)`)

https://github.com/opsani/servox/blob/df5f3bebd733e9a5ee296bc59c03b0ade076877d/servo/telemetry.py#L89

**Pull request merged**
